File: Assets/Scripts/HotUpdate.cs
//using NETCore.Encrypt;
//using System;
//using System.Collections;
//using System.Collections.Generic;
//using System.IO;
//using System.Text;
//using UnityEngine;
//using UnityEngine.Networking;
using System.IO;
using System.Net;
using UnityEngine;
using UnityEngine.Networking;
using UnityEngine.UI;

public class HotUpdate
{
    // Start is called before the first frame update
    public static int mReleaseVersion = 1;
    public static int mMajorVersion = 0;

    public static Slider objProcessBar;
    public static Text Loadingprogress;
    public static Text Loadingimformation;


    public HotUpdate()
    {
        Debug.Log("HotUpdate Class " + mReleaseVersion + "." + mMajorVersion);
        objProcessBar = GameObject.Find("Slider").GetComponent<Slider>();
        Loadingprogress = GameObject.Find("Loadingprogress").GetComponent<Text>();
        Loadingimformation = GameObject.Find("Loadingimformation").GetComponent<Text>();
    }

    public static void ChangeLoadingprogress(string s)
    {
        Loadingprogress.text = "Loading progress   " + s.ToString() + "%";
    }
    public static void ChangeLoadingimformation(string s)
    {
        Loadingimformation.text += s.ToString() + "\n";
    }
    public static void ChangeSlider(string s)
    {
        double process = System.Convert.ToDouble(s);
        objProcessBar.value = (float)process;
    }
    public static void ClearSlider()
    {
        WebUtils.processSlider = (0).ToString();
        WebUtils.processText = (0).ToString();
        new System.Threading.Thread(delegate () { WebUtils.isDone = false; System.Threading.Thread.Sleep(1000); WebUtils.isDone = true; }).Start();
        WebUtils.processSlider = (1).ToString();
        WebUtils.processText = (100).ToString();
    }
    public static void FullSlider()
    {
        //new System.Threading.Thread(delegate () { WebUtils.isDone = false; System.Threading.Thread.Sleep(500); WebUtils.isDone = true; }).Start();
    }

    //    /// <summary>
    //    /// 检查更新
    //    /// </summary>
    //    //void Awake()
    //    //{

    //    //    return;

    //    //    //string packageName = "com." + Application.companyName + "." + Application.productName;
    //    //1、创建版本文件config.ini
    //    CreateVersionFile();
    //    //2、判断是否需要更新
    //    if (!IsUpdate())
    //        return;

    //    //3、判断是否需要释放APK
    //    if (!Directory.Exists(Path.Combine(Application.persistentDataPath, "Lua")))
    //    {
    //        //3.1、从文件管理器中获取APK
    //        GetApkFromFileManager();
    //        //3.2、 解压APK
    //        UnZipTool.UnZipApk(Application.persistentDataPath + "/base.apk", Path.Combine(Application.persistentDataPath, "Lua"));
    //    }

    //    //4、获取服务器压缩文件
    //    GetZipFromServer();
    //    //5、解压第4步中获取的压缩文件
    //    UnZipFile();
    //    //6、还原压缩文件中的差分包或将压缩文件中的Lua文件复制到本地文件
    //    RedutionFile();

    //    //7、 更新本地版本文件version.ini的内容
    //    FileUtils.CreateFile(Path.Combine(Application.persistentDataPath, "config.ini"), WebUtils.GetByteFromServer("version.txt"));
    //    //}

    //    private void RedutionFile()
    //    {
    //        string[] files = Directory.GetFiles(Path.Combine(Application.persistentDataPath, "Temp"));
    //        foreach (string file in files)
    //        {
    //            if (file.EndsWith(".lua"))
    //            {
    //                FileUtils.CopyFileToPath(file, file.Replace("Temp", "Lua"));
    //            }
    //        }
    //        DiffUtils.ReductionFile(Path.Combine(Application.persistentDataPath, "Lua"), Path.Combine(Application.persistentDataPath, "Temp/filelist.txt"));
    //    }

    //    private void UnZipFile()
    //    {
    //        string version = Encoding.UTF8.GetString(FileUtils.ReadFileBytes(Path.Combine(Application.persistentDataPath, "config.ini")));
    //        string[] value = version.Split('=');
    //        value[1] = value[1].Trim();
    //        string serverVersion = Encoding.UTF8.GetString(WebUtils.GetByteFromServer("version.txt"));
    //        string[] serverValue = serverVersion.Split('=');
    //        serverValue[1] = serverValue[1].Trim();

    //        string filename = "v" + value[1] + "-v" + serverValue[1] + ".zip";
    //        Debug.LogError("\n=========\n" + filename + "\n=========\n");
    //        UnZipTool.UnZip(filename, Path.Combine(Application.persistentDataPath, "Temp"));
    //    }

    //    private void GetZipFromServer()
    //    {
    //        string version = Encoding.UTF8.GetString(FileUtils.ReadFileBytes(Path.Combine(Application.persistentDataPath, "config.ini")));
    //        string[] value = version.Split('=');
    //        value[1] = value[1].Trim();
    //        string serverVersion = Encoding.UTF8.GetString(WebUtils.GetByteFromServer("version.txt"));
    //        string[] serverValue = serverVersion.Split('=');
    //        serverValue[1] = serverValue[1].Trim();

    //        string filename = "v" + value[1] + "-v" + serverValue[1] + ".zip";
    //        WebUtils.GetFileFromServer(filename);
    //    }

    //    /// <summary>
    //    /// 判断是否需要更新
    //    /// 通过获取服务器版本文件和本地版本文件进行比较
    //    /// </summary>
    //    /// <returns></returns>
    //    private bool IsUpdate()
    //    {
    //        if (Directory.Exists(Application.persistentDataPath + "/Temp"))
    //            Directory.Delete(Application.persistentDataPath + "/Temp", true);
    //        string serverVersion = Encoding.UTF8.GetString(WebUtils.GetByteFromServer("version.txt"));
    //        string localVersion = Encoding.UTF8.GetString(FileUtils.ReadFileBytes(Path.Combine(Application.persistentDataPath, "config.ini")));
    //        string[] serverInfo = serverVersion.Split('=');
    //        string[] localInfo = localVersion.Split('=');
    //        if (serverInfo[1].Trim() == localInfo[1].Trim())
    //            return false;
    //        return true;
    //    }

    //    /// <summary>
    //    /// 通过UnityWebRequest从服务器端下载文件名为filename的文件
    //    /// </summary>
    //    /// <param name="fileName">需要下载的文件名</param>
    //    /// <returns></returns>
    //    void ReadAssetBundle(string fileName)
    //    {
    //        string url = "http://10.230.17.74/" + fileName;
    //        UnityWebRequest request = UnityWebRequest.Get(url);
    //        request.SendWebRequest();
    //        while (!request.isDone)
    //        {
    //            Debug.Log("ReadAssetBundle:\n====================\n" + request.downloadProgress + "\n====================");

    //        }

    //        if (request.isDone)                  //下载完成
    //        {
    //            byte[] bytes = request.downloadHandler.data;

    //            if (Application.platform == RuntimePlatform.Android)
    //            {
    //                string path = Application.persistentDataPath + "/Android/Lua/";
    //                if (!Directory.Exists(path))
    //                    Directory.CreateDirectory(path);

    //                FileUtils.CreateFile(path + fileName, bytes);

    //            }
    //#if UNITY_EDITOR
    //            UnityEditor.AssetDatabase.Refresh();
    //#endif
    //            Debug.Log(" 下载完成");
    //            //UnZipTool.UnZipPackage(fileName);
    //        }
    //    }

    //    /// <summary>
    //    /// 从手机的文件管理器中，获取apk文件，并放到persistent目录下
    //    /// </summary>
    //    void GetApkFromFileManager()
    //    {
    //        WebUtils.GetApkFromFile();

    //        //string files = Application.streamingAssetsPath.Replace("jar:", "");
    //        //files = files.Replace("!/assets", "");
    //        //using (WWW www = new WWW(files))
    //        //{
    //        //    while (!www.isDone)
    //        //    {
    //        //        Debug.Log("GetApkFromFileManager:\n=================\n" + www.progress + "\n=================\n");
    //        //    }

    //        //    Debug.Log("\n==================\n==================\n" + Application.streamingAssetsPath + "\n/data/app/" + "\n==================\n==================\n");
    //        //    if (www.isDone)
    //        //    {
    //        //        Debug.Log("\n==================\n==================\n" + Application.streamingAssetsPath + "\n" + files + "\n==================\n==================\n");
    //        //        FileStream output = new FileStream(Application.persistentDataPath + "/base.apk", FileMode.Create, FileAccess.Write);
    //        //        byte[] buffer = www.bytes;
    //        //        //APKMD5 = EncryptProvider.Md5(Encoding.UTF8.GetString(buffer));
    //        //        Debug.Log("\n=======================\n" + APKMD5 + "\n=======================\n");
    //        //        output.Write(buffer, 0, buffer.Length);
    //        //        output.Close();
    //        //        output.Dispose();
    //        //    }
    //        //}
    //    }

    //    void CreateVersionFile()
    //    {
    //        string filename = Path.Combine(Application.persistentDataPath, "config.ini");
    //        if (File.Exists(filename))
    //            return;
    //        string content = "version = " + mReleaseVersion + "." + mMajorVersion;
    //        byte[] buffer = Encoding.UTF8.GetBytes(content);
    //        FileUtils.CreateFile(filename, buffer);
    //    }

}

File: Assets/Lua/HotUpdate.lua
HotUpdate = {}
local this = HotUpdate
local cor = coroutine.create(
    function() 
        HotUpdate.HotUpdateProcess() 
    end
    );

function this.Awake(object)
    coroutine.resume(cor);
    -- coroutine.resume(cor);
end

function this.Update()
    if WebUtils.isDone then
        WebUtils.isDone = false;
        coroutine.resume(cor);
    end
end

function this.HotUpdateProcess()
    HotUpdateClass.ChangeLoadingimformation("检查更新中");
    HotUpdateClass.ClearSlider()
    -- HotUpdateClass.FullSlider()
    if not WebUtils.isDone then
        coroutine.yield()
    end
    -- 1、创建版本文件config.ini
    this.CreateVersionFile();
    -- 2、判断是否需要更新
    if not this.IsUpdate() then
        HotUpdateClass.ChangeLoadingimformation("目前为最新版本\n正在进入游戏");
        HotUpdateClass.ClearSlider()
        -- HotUpdateClass.FullSlider()
        if not WebUtils.isDone then
            coroutine.yield()
        end
        SceneManagement.SceneManager.LoadScene("MainMenu")
        return;
    end
    HotUpdateClass.ChangeLoadingimformation("检测本地资源");
    HotUpdateClass.ClearSlider()
    -- HotUpdateClass.FullSlider()
    if not WebUtils.isDone then
        coroutine.yield()
    end
    -- 3、判断是否需要释放APK
    if not FileUtils.DirectoryExists(Path.Combine(Application.persistentDataPath, "Lua")) then
        -- 3.1、从文件管理器中获取APK
        HotUpdateClass.ChangeLoadingimformation("获取本地资源");
        this.GetApkFromFileManager();
        if not WebUtils.isDone then
            coroutine.yield()
        end
        -- 3.2、 解压APK
        HotUpdateClass.ChangeLoadingimformation("释放本地资源");
        UnZipTool.UnZipApk(Application.persistentDataPath .. "/base.apk", Path.Combine(Application.persistentDataPath, "Lua"));
    end
    HotUpdateClass.ChangeLoadingimformation("本地资源加载完成");
    HotUpdateClass.ClearSlider()
    -- HotUpdateClass.FullSlider()
    if not WebUtils.isDone then
        coroutine.yield()
    end

    -- 4、获取服务器压缩文件
    HotUpdateClass.ChangeLoadingimformation("正在获取服务器端资源");
    this.GetZipFromServer();
    if not WebUtils.isDone then
        coroutine.yield()
    end

    -- 5、解压第4步中获取的压缩文件
    HotUpdateClass.ChangeLoadingimformation("释放服务器端资源");
    this.UnZipFile();
    HotUpdateClass.ClearSlider()
    -- HotUpdateClass.FullSlider()
    if not WebUtils.isDone then
        coroutine.yield()
    end
    -- 6、还原压缩文件中的差分包或将压缩文件中的Lua文件复制到本地文件
    HotUpdateClass.ChangeLoadingimformation("更新本地资源中");
    this.RedutionFile();
    HotUpdateClass.ClearSlider()
    -- HotUpdateClass.FullSlider()
    if not WebUtils.isDone then
        coroutine.yield()
    end

    -- 7、 更新本地版本文件version.ini的内容
    FileUtils.CreateFile(Path.Combine(Application.persistentDataPath, "config.ini"), WebUtils.GetByteFromServer("version.txt"));
    HotUpdateClass.ChangeLoadingimformation("更新完成");
    HotUpdateClass.ClearSlider()
    -- HotUpdateClass.FullSlider()
    if not WebUtils.isDone then
        coroutine.yield()
    end
    SceneManagement.SceneManager.LoadScene("MainMenu")
end

function this.CreateVersionFile()
    local filename = Path.Combine(Application.persistentDataPath, "config.ini");
    if FileUtils.FileExists(filename) then
        return;
    end
    FileUtils.CreateFile(filename, Encoding.GetBytes("version = " .. HotUpdateClass.mReleaseVersion .. "." .. HotUpdateClass.mMajorVersion));
end

function this.IsUpdate()
    if FileUtils.DirectoryExists(Application.persistentDataPath .. "/Temp") then
        FileUtils.DirectoryDelete(Application.persistentDataPath .. "/Temp", true);
    end
    local serverVersion = Encoding.GetString(WebUtils.GetByteFromServer("version.txt"));
    local localVersion = Encoding.GetString(FileUtils.ReadFileBytes(Path.Combine(Application.persistentDataPath, "config.ini")));
    local serverInfo = Split(serverVersion, '=');
    local localInfo = Split(localVersion, '=');
    if Trim(serverInfo[2]) == Trim(localInfo[2]) then
        return false;
    end
    return true;
end

function this.GetApkFromFileManager()
    WebUtils.GetApkFromFile();
end

function this.GetZipFromServer()
    local version = Encoding.GetString(FileUtils.ReadFileBytes(Path.Combine(Application.persistentDataPath, "config.ini")));
    local value = Split(version, '=');
    value[2] = Trim(value[2]);
    local serverVersion = Encoding.GetString(WebUtils.GetByteFromServer("version.txt"));
    local serverValue = Split(serverVersion, '=');
    serverValue[2] = Trim(serverValue[2]);

    
    local filename = "v" .. value[2] .. "-v".. serverValue[2] .. ".zip";
    WebUtils.GetFileFromServer(filename);
end

function this.UnZipFile()
    local version = Encoding.GetString(FileUtils.ReadFileBytes(Path.Combine(Application.persistentDataPath, "config.ini")));
    local value = Split(version, '=');
    value[2] = Trim(value[2]);
    local serverVersion = Encoding.GetString(WebUtils.GetByteFromServer("version.txt"));
    local serverValue = Split(serverVersion, '=');
    serverValue[2] = Trim(serverValue[2]);

    local filename = "v" .. value[2] .. "-v" .. serverValue[2] .. ".zip";
    UnZipTool.UnZip(filename, Path.Combine(Application.persistentDataPath, "Temp"));
end

function this.RedutionFile()
    local files = FileUtils.DirectoryGetFiles(Path.Combine(Application.persistentDataPath, "Temp"));
    Debugger.Log(files.Length)
    local i = 0
    while i < files.Length do
        if EndWith(files[i], ".lua") then
            local t = string.gsub(files[i],"Temp","Lua",1)
            FileUtils.CopyFileToPath(files[i], t);
        end
        i = i + 1
    end
    -- foreach (string file in files)
    -- if (file.EndsWith(".lua"))
    DiffUtils.ReductionFile(Path.Combine(Application.persistentDataPath, "Lua"), Path.Combine(Application.persistentDataPath, "Temp/filelist.txt"));
end

function Split(szFullString, szSeparator)
    local nFindStartIndex = 1
    local nSplitIndex = 1
    local nSplitArray = {}
    while true do
       local nFindLastIndex = string.find(szFullString, szSeparator, nFindStartIndex)
       if not nFindLastIndex then
        nSplitArray[nSplitIndex] = string.sub(szFullString, nFindStartIndex, string.len(szFullString))
        break
       end
       nSplitArray[nSplitIndex] = string.sub(szFullString, nFindStartIndex, nFindLastIndex - 1)
       nFindStartIndex = nFindLastIndex + string.len(szSeparator)
       nSplitIndex = nSplitIndex + 1
    end
    return nSplitArray
end

function Trim(s)
   return s:match "^%s*(.-)%s*$"
end

function EndWith(str, substr)  
    if str == nil or substr == nil then  
        return nil, "the string or the sub-string parameter is nil"  
    end  
    local str_tmp = string.reverse(str)  
    local substr_tmp = string.reverse(substr)  
    if string.find(str_tmp, substr_tmp) ~= 1 then  
        return false  
    else  
        return true  
    end  
end